classes/order/OrderDetail.php

```diff
@@ -536,25 +536,38 @@ protected function setVirtualProductInformation($product)
      * Check the order status.
      *
      * @param array $product
-     * @param int $id_order_state
+     * @param int $orderStateId
      */
-    protected function checkProductStock($product, $id_order_state)
+    protected function checkProductStock($product, $orderStateId)
     {
-        if ($id_order_state != Configuration::get('PS_OS_CANCELED') && $id_order_state != Configuration::get('PS_OS_ERROR')) {
-            $update_quantity = true;
-            if (!StockAvailable::dependsOnStock($product['id_product'])) {
-                $update_quantity = StockAvailable::updateQuantity($product['id_product'], $product['id_product_attribute'], -(int) $product['cart_quantity'], $product['id_shop'], true);
-            }
-
-            if ($update_quantity) {
-                $product['stock_quantity'] -= $product['cart_quantity'];
-            }
-
-            if ($product['stock_quantity'] < 0 && Configuration::get('PS_STOCK_MANAGEMENT')) {
-                $this->outOfStock = true;
-            }
-            Product::updateDefaultAttribute($product['id_product']);
+        $dismissOrderStateIds = Configuration::getMultiple([
+            'PS_OS_CANCELED',
+            'PS_OS_ERROR',
+        ]);
+        if (in_array($orderStateId, $dismissOrderStateIds)) {
+            return;
+        }
+        $quantityUpdated = true;
+
+        if (!StockAvailable::dependsOnStock($product['id_product'])) {
+            $quantityUpdated = StockAvailable::updateQuantity(
+                $product['id_product'],
+                $product['id_product_attribute'],
+                -(int) $product['cart_quantity'],
+                $product['id_shop'],
+                $orderStateId == Configuration::get('PS_OS_DELIVERED'),
+                [
+                    'id_order' => $this->id_order,
```

Comment:
The stock movement type is missing from the parameters in this method call. Logically, any stock movement should include an `id_stock_mvt_reason` (see https://github.com/PrestaShop/PrestaShop/blob/develop/src/Core/Stock/StockManager.php#L396-L398) so that the merchant can later tell why the stock was changed.

However, there is no stock movement type for "order shipped" as we can see in ps_configuration:
<img width="703" alt="Screenshot 2022-02-09 at 16 45 54" src="https://user-images.githubusercontent.com/1009343/153278497-9308ee76-bc0a-4819-8d7c-52799a73c0be.png">

It's not clear to me whether this was forgotten or done in purpose. According to the original spec, tracking physical and available quantities separately is out of scope, so following this logic (we only track available quantities), the available stock should be decremented when the order is created, not when the order is shipped. This would be consistent with the "order shipped" stock movement type not existing.

Please advise @MatShir @fabienvallon

Reply:
@eternoendless Agreed.

I got it wrong when I assumed the default movement reason was `PS_STOCK_MVT_DEC_EMPLOYEE_EDITION`, but it's a quick fix.
A point remains unclear. An order created from BO should get treated as an usual customer order, or as an employee edition ?
I suppose the "Customer Order" reason includes the "order shipped" state, which has no impact on quantity.

Unfortunately the current behavior in FO is inconsistent with physical and available quantity tracking:
- a new order adds no stock movement until delivered
- movement is created only when order status is switched to delivered
- movement reason ID is 3 => `PS_STOCK_MVT_REASON_DEFAULT` or `PS_STOCK_CUSTOMER_ORDER_REASON`

And the current behavior in BO (before this PR) is messed up as well:
- a new order adds a new movement, whatever the order status, without order reference
- switching order status to "Delivered" duplicates the same movement with correct order reference

As you mentioned, tracking available quantity requires a new movement immediately when the order is created.

Should we keep FO behaviour ? or fix both FO and BO to track available quantity accurately ?

Reply:
@eternoendless IMHO the easiest way to track both available and physical quantities is to split customer order reason in 2 reasons :
- Customer Order Reserve => impacts available quantity only when order is created
- Customer Order Shipping => impact physical quantity only when order status is updated to shipping

Then sum all movements:
- available tracking should exclude `shipping` movement reasons
- physical tracking should exclude `reserve`  movement reasons

Reply:
Actually the technical solution I provided is out of the scope described in original specifications : https://forge.prestashop.com/browse/FF-1

A stock movement is supposed to track physical quantities only, i.e. stock quantities physically present in a given warehouse. A movement was never intended to track stock quantities available to order.

Even if the solution I provided could work in theory, it's certainly over complicated to implement 2 distinct business needs at the same technical place.
Then I propose to split the technical implementation in 2 : 
- keep stock movements to track physical quantities only
- add a new table to track available quantities only : `stock_reserve`

Then the physical quantity is easy to obtain from the stock movements table alone.
And the available quantity requires to query the stock table joined to both tables : movements & reserves

IMHO opinion this implementaiton is safer, and it allows all kind of business behaviors according to shop configuration.
Example : to deactivate the available quantity treshold, don't join stock reserves or don't insert any reserved quantity at all.

Reply:
We need to discuss this, but I believe `StockAvailable.reserved_quantity` is already here for that purpose

Reply:
@jolelievre 
I suppose you refer the table column as `StockAvailable.reserved_quantity` ? class `StockAvailable` owns no `reserved_quantity` property.
Looks like indeed that this column intents to keep track of the total quantity of reserved products in stock. Which is not exactly the same as historization tracking feature alike stock movements. If the product doesn't need to historize reservations and their cancellation, sounds good enough to me :)